src/ck-core/ckrdmadevice.C

```diff
@@ -0,0 +1,300 @@
+  int pool_size = CsvAccess(gpu_manager).cuda_ipc_event_pool_size;
+  int device_index = dm->global_index;
+  cuda_ipc_device_info& my_device_info = CsvAccess(gpu_manager).cuda_ipc_device_infos[device_index];
+
+  // Free IPC events that are complete
+  // TODO: Don't do this every time but only when the event pool is somewhat empty
+  for (int i = 0; i < pool_size; i++) {
+    int& event_flag = my_device_info.event_pool_flags[i];
+    cudaEvent_t& ev = my_device_info.dst_event_pool[i];
+    size_t& buff_offset = my_device_info.event_pool_buff_offsets[i];
+    // For a used event, check if it's complete and mark as free if so
+    if (event_flag != 0) {
+      // Check in shared memory if receiver has invoked the memcpy from
+      // the device comm buffer on sender to destination buffer
+      cuda_ipc_event_shared* shm_event_shared =
+        (cuda_ipc_event_shared*)((char*)CsvAccess(gpu_manager).shm_ptr
+            + CsvAccess(gpu_manager).shm_chunk_size * device_index
+            + sizeof(cudaIpcMemHandle_t)) + i;
+      bool can_query = false;
+      pthread_mutex_lock(&shm_event_shared->lock);
+      if (shm_event_shared->dst_flag == true) {
+        shm_event_shared->dst_flag = false;
+        can_query = true;
+      }
+      pthread_mutex_unlock(&shm_event_shared->lock);
+
+      // If the receiver has invoked the memcpy,
+      // the sender can query the event for completion
+      if (can_query) {
+        if (cudaEventQuery(ev) == cudaSuccess) {
+          // Event completion means that the transfer from source device comm buffer
+          // to dest buffer is complete, so free the allocated block
+          if (event_flag == 1) {
+            dm->free_comm_buffer(buff_offset);
+          }
+
+          // Mark event as free
+          event_flag = 0;
+        }
+      }
+    }
+  }
+
+  // Allocate CUDA IPC events from the pool
+  // Two events are used per message:
+  // 1) Recorded by the sender after 'source buffer -> device comm buffer' cudaMemcpy.
+  //    Can be used by the sender to determine if the sender buffer is free for reuse.
+  //    It is also used by the receiver to create a dependency for the second cudaMemcpy
+  //    ('device comm buffer -> dest buffer')
+  // 2) Recorded by the receiver after 'device comm buffer -> dest buffer' cudaMemcpy.
+  //    It is used by the sender to determine when the allocated block on
+  //    device comm buffer and IPC events can be freed.
+  for (int i = 0; i < pool_size; i++) {
+    int& event_flag = my_device_info.event_pool_flags[i];
+    size_t& buff_offset = my_device_info.event_pool_buff_offsets[i];
+    if (event_flag == 0) {
+      event_flag = 1;
+      buff_offset = comm_offset;
+      return i;
+    }
+  }
+
+  return -1;
+}
+
+// Performs sender-side operations necessary for device zerocopy
+void CkRdmaDeviceOnSender(int dest_pe, int numops, CkDeviceBuffer** buffers) {
+  // TODO: Need to handle the case where the destination PE could be wrong
```

Comment:
Are there any checks on this at runtime? Shouldn't we abort or warn users about this?

Reply:
There isn't one currently, but I'll add a check on the receiver side to compare the current PE and the destination PE and abort if they don't match. As of now the user is expected to build charm with `-DCMK_GLOBAL_LOCATION_UPDATE` (which will go into the documentation).